algorithm-simulator/src/main/scala/style95/scaler/QueuingTheory.scala

```diff
@@ -0,0 +1,117 @@
+package style95.scaler
+
+import akka.actor.ActorRef
+import style95.StatusLogger.PredictedTps
+
+import scala.collection.immutable.Queue
+import scala.concurrent.duration._
+import scala.math.{ceil, max}
+
+// Time series analysis + queuing theory + ad hoc rules
+// Algorithm outline:
+//   1. Use single exponential smoothing to forecast feature load.
+//   2. Scaling out with queuing theory (M/M/1 queue) to meet schedule latency requirements.
+//   3. Scaling in with ad hoc rules
+class QueuingTheory(
+    creatingHint: FiniteDuration, // a hint value for how long it takes to create a new container
+    consultInterval: FiniteDuration, // the time interval between successive invocations of the scaler
+    logger: ActorRef, // log status to file
+    alpha: Double, // parameter of time series analysis
+    slo: FiniteDuration, // schedule latency service level objective
+    idleTimeout: FiniteDuration, // wKZait how long before killing an idle container
+) extends Scaler {
+  require(slo > 0.second, "latency requirement must be greater than zero")
+
+  private var s = 0D
+  // A container buffer is maintained to make sure an idle container won't be killed
+  // prematurely before the idleTimeout is expired.
+  private var idleBuffer = Queue.empty[(ActorRef, Long)]
+
+  override def decide(info: DecisionInfo): Decision = {
```

Comment:
Could you align this `decide` function with Queueing theory for me?

Reply:
No problem, but please forgive me, I need to understand it better before I present it to you.

This version of scaler is still flawed. It's too responsive to request rate fluctuation and the queuing model is inaccurate, which cause observable over-provisioning.

 It needs more iteration. But I opened a RP anyway because there are already many changes 😄 .

Reply:
Got it!